### PR TITLE
Move the verification of iamges part to the backend

### PR DESCRIPTION
**Pull request by nilxam**
Move verify the images to the backend and also use needle:tags in the  backend. verify_sound_image in basetest.pm just to do post-work.

### CODE REVIEW

**Comment by coolo** [comment]
But you do more - you also create the $backend variable and leak it into the autotest process.

**Comment by nilxam** [comment]
indeed. updated.

**Comment by coolo** [comment]
you didn't really motivate this change though - why are you touching it?

**Comment by nilxam** [comment]
a lazy way to fix firefox_audio missing needle issue, and I wondering needle::tag() call really works currently, looks like needle::all gives 0.

note. my instance has patched a lot of different place though...

**Comment by coolo** [comment]
wait a second. this means the firefox needle is compared in the autotest process? We need to fix that. Only the backend reloads the needles. So instead of working around it, make sure verify_sound_image is actually working on filenames passed to the backend.

**Comment by nilxam** [comment]
knock! knock!

**Comment by okurz** [comment]
can you fix the typo s/iamges/images/ in the commit message and please describe a bit more what this is intended to do? You are describing what the change is doing but not why and how this fixes the firefox sound test I guess? Can you provide a link to a local verification run?

File: backend/baseclass.pm [hunk 1]
```diff
@@ -900,6 +900,22 @@ sub last_screenshot_name {
...
+    my $mustmatch = $args->{mustmatch};
+
+    my $img = tinycv::read($imgpath);
+    my $needles = needle::tags($mustmatch) || [];
+
+    my ($foundneedle, $failed_candidates) = $img->search($needles, 0, 1);
+    if ($foundneedle) {
+        return {imgpath => $imgpath, found => $foundneedle, candidates => $failed_candidates};
+    }
+    return {imgpath => $imgpath, candidates => $failed_candidates};
```
Comment (aaannz) [inline backend/baseclass.pm]: I don't get why are you returning the same `$imgpath`? I'm probably missing the whole context, but looks weird.

File: basetest.pm [hunk 2]
```diff
@@ -127,10 +127,12 @@ sub record_screenmatch {
...
-    if ($match->{needle}->has_property('workaround')) {
-        $result->{dent} = 1;
-        $self->{dents}++;
-        bmwqemu::diag("needle '$h->{name}' is a workaround");
+    for my $t (@${properties}) {
+        if ($t eq 'workaround') {
+            $result->{dent} = 1;
+            $self->{dents}++;
+            bmwqemu::diag("needle '$h->{name}' is a workaround");
+        }
```
Comment (aaannz) [inline basetest.pm]: What is the reason for this change? `$match->{needle}` is no longer an object?

File: backend/baseclass.pm (hunk 1 above)
Comment (nilxam) [inline backend/baseclass.pm]: ah, yes, it's superfluous.

File: basetest.pm (hunk 2 above)
Comment (nilxam) [inline basetest.pm]: bless/unbless problem, I can bless it if preferred.

**Comment by nilxam** [comment]
http://147.2.211.60/tests/382 local verification run

File: isotovideo [hunk 3]
```diff
@@ -349,6 +349,19 @@ while ($loop) {
                     }})->{tags};
             next;
         }
+        if ($rsp->{cmd} eq 'verify_sound_image') {
```
Comment (coolo) [inline isotovideo]: please note. Just prefix the query with backend_ and let isotovideo pass it as is

Comment (coolo) [inline isotovideo]: aehm, please not

**Comment by nilxam** [comment]
updated. and http://147.2.211.60/tests/386 for local verification run

**Comment by coolo** [comment]
perfecto!

**Comment by okurz** [comment]
awesome, thank you
